libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -251,7 +251,9 @@ string IRGenerator::dispatchRoutine(ContractDefinition const& _contract)
 	Whiskers t(R"X(
 		if iszero(lt(calldatasize(), 4))
 		{
-			let selector := <shr224>(calldataload(0))
+			mstore(0, 0)
+			calldatacopy(28, 0, 4)
+			let selector := mload(0)
```

Comment:
While this is "less gas" for wasm output, I think it costs more in EVM. Do we want to merge it this way?

Reply:
I wish we already had https://github.com/ethereum/solidity/issues/5107

Reply:
@axic @chriseth if the evm version is more costly after doing this, would it make sense to generate different code here, for evm vs. ewasm case?

Reply:
Can check what the difference is exactly?

Reply:
@chriseth @axic I did some measurements.  I ran `soltest` on all `semanticTests` and calculated the gas-use difference with and without the code-change defined here. The gas-use does not change for `42%` of the calls done by `semanticTests`, `22%` of all calls have an additional gas-use of `18 gas`. Creation calls get cheaper. 

<img width="942" alt="diagram-measurements" src="https://user-images.githubusercontent.com/5008794/84523057-4c68c900-ac9d-11ea-8e6e-53b000665d8b.png">

Reply:
It should be 26 gas more expensive unless we can optimize the initial `mstore(0,0)` away, in which case it would be 17 gas.

Reply:
I was surprised that this change had impact on so many calls differently. I expected to see that some calls get more expensive by a given constant. 

Probably I did something wrong during the measurements, basically I ran `soltest` with `--show-messages` with and without this change. And logged the output into two files, after this I was just grepping for `gas used:`. Then I calculated the difference of these two files per line. I thought that the tests are executed deterministically, but probably exactly here is my mistake. Do you think the graphs are wrong?

Reply:
Yeah, especially the max and min values look weird...

Reply:
It also looked weird to me, but I checked now everything again. The graphs should be correct.

I uploaded the different logs `original` and `changed` here: https://www.dropbox.com/sh/8brotd0m3q98o23/AABlIjg38YpNyBqwSSO1PeQ3a?dl=0

I used this program to parse the logs: https://gist.github.com/aarlt/8cbf14135cc361827fe52dc6f70bac35 the output of that program running on the given logs is also included there.